### PR TITLE
Feature/redis

### PR DESCRIPTION
**Pull request by ajax-makarytskyi-o**
Added Redis to project.

### CODE REVIEW

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/config/RedisConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+class RedisConfiguration {
+
+    @Value("\${redis.host}")
+    private lateinit var host: String
+
+    @Value("\${redis.port}")
+    private lateinit var port: String
+
+    @Value("\${redis.timeout}")
+    private lateinit var timeout: String
```
Comment (ajax-marchenko-s) [inline domainservice/src/main/kotlin/com/makarytskyi/rentcar/config/RedisConfiguration.kt]: you can encapsulate these properties in data class

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/service/impl/CarServiceImpl.kt [hunk 2]
```diff
@@ -15,7 +15,9 @@ import reactor.kotlin.core.publisher.switchIfEmpty
 
 @InvocationTracker
 @Service
-internal class CarServiceImpl(private val repository: CarRepository) : CarService {
+internal class CarServiceImpl(
+    private val repository: CarRepository
```
Comment (ajax-marchenko-s) [inline domainservice/src/main/kotlin/com/makarytskyi/rentcar/service/impl/CarServiceImpl.kt]: add comma

File: domainservice/src/test/kotlin/com/makarytskyi/rentcar/repository/MongoCarRepositoryTest.kt [hunk 3]
```diff
@@ -13,11 +13,13 @@ import org.junit.jupiter.api.Assertions.assertEquals
...
 import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Qualifier
 import reactor.kotlin.test.test
 
-internal class CarRepositoryTest : ContainerBase {
+internal class MongoCarRepositoryTest : ContainerBase {
 
     @Autowired
+    @Qualifier("mongoCarRepository")
     lateinit var carRepository: CarRepository
```
Comment (ajax-marchenko-s) [inline domainservice/src/test/kotlin/com/makarytskyi/rentcar/repository/MongoCarRepositoryTest.kt]: make it private

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/config/RedisConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.data.redis.connection.RedisStandaloneConfiguration
+import org.springframework.data.redis.connection.lettuce.LettuceClientConfiguration
+import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory
+import org.springframework.data.redis.core.ReactiveRedisTemplate
+import org.springframework.data.redis.serializer.RedisSerializationContext
+import org.springframework.data.redis.serializer.RedisSerializer
+import org.springframework.data.redis.serializer.StringRedisSerializer
+
+@Configuration
+@EnableCaching
```
Comment (ajax-yurkovskyi-t) [inline domainservice/src/main/kotlin/com/makarytskyi/rentcar/config/RedisConfiguration.kt]: why u need this?

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import org.springframework.stereotype.Repository
+import reactor.core.publisher.Mono
+import reactor.core.scheduler.Schedulers
+import reactor.kotlin.core.publisher.switchIfEmpty
+import reactor.util.retry.Retry
+
+@Repository
+@Primary
+internal class RedisCarRepository(
+    @Qualifier("mongoCarRepository")
```
Comment (ajax-marchenko-s) [inline domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt]: do you need this qualifier?

**Review by ajax-marchenko-s — COMMENTED** [review]
Great work!

File: domainservice/src/test/kotlin/com/makarytskyi/rentcar/repository/RedisCarRepositoryTest.kt [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                    .test()
+                    .expectNext(true)
+                    .verifyComplete()
+            }
+
+        redisTemplate.hasKey(plateRedisKey(car.plate!!))
+            .test()
+            .expectNext(true)
+            .verifyComplete()
+    }
```
Comment (ajax-yurkovskyi-t) [inline domainservice/src/test/kotlin/com/makarytskyi/rentcar/repository/RedisCarRepositoryTest.kt]: why u dont check here what u saved to redis

File: domainservice/src/test/kotlin/com/makarytskyi/rentcar/repository/RedisCarRepositoryTest.kt [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    lateinit var mapper: ObjectMapper
+
+    @Test
+    fun `redis repository should cache the car after calling getById`() {
+        // GIVEN
+        val car = mongoCarRepository.create(randomCar()).block()!!
+        val carId = car.id.toString()
+
+        // WHEN
+        redisCarRepository.findById(carId).block()
```
Comment (ajax-yurkovskyi-t) [inline domainservice/src/test/kotlin/com/makarytskyi/rentcar/repository/RedisCarRepositoryTest.kt]: i think u can reuse it in then block

**Review by ajax-yurkovskyi-t — COMMENTED** [review]
good job

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    private val redisProperties: RedisProperties,
+) : CarRepository by mongoCarRepository {
+
+    override fun findById(id: String): Mono<MongoCar> =
+        reactiveRedisTemplate.opsForValue().get(idRedisKey(id))
+            .map { mapper.readValue(it, MongoCar::class.java) }
+            .onErrorResume { Mono.empty() }
+            .switchIfEmpty {
+                mongoCarRepository.findById(id)
+                    .subscribeOn(Schedulers.boundedElastic())
```
Comment (ajax-shulepa-a) [inline domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt]: why do you need subscribeOn here?

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    private val mapper: ObjectMapper,
+    private val redisProperties: RedisProperties,
+) : CarRepository by mongoCarRepository {
+
+    override fun findById(id: String): Mono<MongoCar> =
+        reactiveRedisTemplate.opsForValue().get(idRedisKey(id))
+            .map { mapper.readValue(it, MongoCar::class.java) }
+            .onErrorResume { Mono.empty() }
+            .switchIfEmpty {
+                mongoCarRepository.findById(id)
```
Comment (ajax-shulepa-a) [inline domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt]: it's better to extract code from switchIfEmpty to a separate function

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            }
+
+
+    override fun findByPlate(plate: String): Mono<MongoCar> =
+        reactiveRedisTemplate.opsForValue().get(plateRedisKey(plate))
+            .map { mapper.readValue(it, MongoCar::class.java) }
+            .onErrorResume { Mono.empty() }
+            .switchIfEmpty {
+                mongoCarRepository.findByPlate(plate)
+                    .publishOn(Schedulers.boundedElastic())
```
Comment (ajax-shulepa-a) [inline domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt]: here I see piblishOn. in findById in the same place - subscribeOn. why?

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt (hunk 8 above)
Comment (ajax-makarytskyi-o) [reply domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt]: I forgot to remove it from here, after I moved `subscribeOn` to `setRedisKey`.

File: domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt (hunk 10 above)
Comment (ajax-makarytskyi-o) [reply domainservice/src/main/kotlin/com/makarytskyi/rentcar/repository/impl/RedisCarRepository.kt]: The same thing, I will remove it. I used 'publishOn' here to check if performance will be better and forgot to remove it.

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**